Stratis.Bitcoin/Utilities/CircularArray.cs

```diff
@@ -0,0 +1,110 @@
+{
+    /// <summary>
+    /// Generic circular array is a fixed length array which stores collection 
+    /// of items. Once the array is full, adding a new item removes the oldest entry. 
+    /// </summary>
+    /// <typeparam name="T">Type of the items stored in the array.</typeparam>
+    /// <remarks>
+    /// Complexity of supported operations:
+    /// <list type="bullet">
+    /// <item><see cref="Add"/> - O(1),</item>
+    /// <item><see cref="AddNoSet"/> - O(1),</item>
+    /// <item><see cref="this[int]"/> - O(1).</item>
+    /// </list>
+    /// </remarks>
+    public class CircularArray<T> : IEnumerable<T> where T : new()
+    {
+        /// <summary>Maximal number of items that can be stored in <see cref="items"/> array.</summary>
+        public int Capacity { get; private set; }
+
+        /// <summary>Number of valid items in <see cref="items"/> array.</summary>
+        public int Count { get; private set; }
+
+        /// <summary>Index in <see cref="items"/> array where the next item will be stored.</summary>
+        public int Index { get; private set; }
+
+        /// <summary>Circular array of items, which holds <see cref="Count"/> valid items, and can store up to <see cref="Capacity"/> items.</summary>
+        private T[] items;
+
+        /// <summary>
+        /// Initializes a new instance of the object.
+        /// </summary>
+        /// <param name="capacity">Maximal number of items that can be stored in the circular array.</param>
+        /// <param name="preinitializeItems">If set to <c>true</c>, all items in the array will be initialized using their default constructor. 
+        /// This can be used to prevent further allocations when the structure is used.</param>
+        public CircularArray(int capacity, bool preinitializeItems = true)
+        {
+            this.Index = 0;
+            this.Count = 0;
+            this.Capacity = capacity;
+            this.items = new T[capacity];
+            if (preinitializeItems)
+            {
+                for (int i = 0; i < capacity; i++)
+                    this.items[i] = new T();
+            }
+        }
+
+        /// <summary>
+        /// Add a new item to the circular array.
+        /// </summary>
+        /// <param name="item">Item to add.</param>
+        /// <remarks>If the array already reached its capacity, this method will replace the oldest item with the new item.</remarks>
+        public void Add(T item)
+        {
+            int index = this.AddNoSet();
+            this.items[index] = item;
+        }
+
+        /// <summary>
+        /// Add a new item to the circular array without setting its value.
+        /// </summary>
+        /// <returns>Index to the array where the new item should be placed.</returns>
+        /// <remarks>This method only moves the array indexes as if a new item was added, but the caller is responsible for actually setting new values of the item.</remarks>
+        public int AddNoSet()
```

Comment:
This should probably be made private. The consumer of the class doesn't need to know at which index the next item be inserted. Also, the name AddNoSet doesn't sound right. Is it really adding an item to the array?